torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -848,20 +894,32 @@ class ShapePropagator {
         [this](Node* node) -> type_vec_t {
           if (auto maybe_tensor_types =
                   gatherTensorTypes<DimensionedTensorType>(node)) {
-            return {broadcast(*maybe_tensor_types, 0)};
+            return {broadcast(*maybe_tensor_types, (*maybe_tensor_types)[0]->scalarType())};
           }
           return {};
         }};
 
-    // NB: we always take the scalar type of the Tensor
-    static const register_formula_for broadcasting_tensor_scalar_ops{
+    static const register_formula_for broadcasting_tensor_scalar_ops_arithmetic{
         {
 
             // Tensor-Scalar operators
             "aten::add(Tensor self, Scalar other, Scalar alpha) -> Tensor",
             "aten::sub(Tensor self, Scalar other, Scalar alpha) -> Tensor",
             "aten::mul(Tensor self, Scalar other) -> Tensor",
             "aten::div(Tensor self, Scalar other) -> Tensor",
+          },
+          [this](Node* node) -> type_vec_t {
+            if (auto maybe_tensor_types =
+                    gatherTensorTypes<DimensionedTensorType>(node)) {
+              return {broadcast(*maybe_tensor_types, getScalarType(node))};
```

Comment:
Same comment here ^^ , give up if alpha is ScalarType

Reply:
Same thing here - for `add` and `sub` here. 

If `other` is an int or float it changes the output type, so you can't resolve the output type if you don't know its concrete type